### DISCUSSION TITLE
Allow creating full stand-alone C# programs in a single .cs

### DISCUSSION THREAD

**Discussion by olfek** [dotnet/csharplang#7552] [Language Ideas] [answered]
Reopening https://github.com/dotnet/csharplang/issues/3502 by @mythz here

Idea 💡
https://github.com/dotnet/csharplang/issues/3502#issuecomment-1732516058

> Be practical everyone ...
> 
> We want:
> ```
> dotnet run main.cs
> ```
> 
> **Without** needing to specify any package references.
> 
>  1. dotnet in a script mode parses the `using` statements and looks for relevant packages in the global `.nuget` packages folder.
>  1. dotnet passes `main.cs` to the Roslyn compiler along with references to the relevant packages found in the global `.nuget` packages folder.
>  1. `dotnet` program runs executable emitted by Roslyn compiler.
>  1. Profit

**Comment by bernd5** [comment]
It is not actually a language but a tooling thing.
Maybe the runtime or sdk repo would be better

**Reply by olfek** [reply]
can this be transferred there?

**Reply by olfek** [reply]
https://github.com/dotnet/sdk

**Reply by olfek** [reply]
@marcpopMSFT ⬆️⬆️⬆️

**Reply by marcpopMSFT** [reply]
We don't use discussions in the SDK repo so I can't just transfer. I know that this suggestion has come up before but I can't seem to find a related issue. CC @richlander and @baronfel in case they are aware of a tracking issue or could write one up in the SDK repo.

**Comment by TahirAhmadov** [comment]
https://www.cs-script.net/

**Comment by marcospgp** [comment]
Looking into this now and it seems like there are ways to run C# standalone scripts, at least according to ChatGPT (not incredibly official). Is there documentation available on this maybe?

**Reply by colejohnson66** [reply]
AFAIK, there are no official plans to add scripting support to C#. The best we can get now is top-level statements in a `Program.cs` file. If you want scripting, [CS-Script](https://www.cs-script.net/) exists. [LINQPad](https://www.linqpad.net/) is also pretty neat. It can handle single expressions all the way to full programs. It does come with a pretty hefty cost, however. Lastly, for a "REPL", [dotnet-script](https://github.com/dotnet-script/dotnet-script) is great.

**Reply by sharpchen** [reply]
> for a "REPL", [dotnet-script](https://github.com/dotnet-script/dotnet-script) is great.

Consider trying out [CSharpRepl](https://github.com/waf/CSharpRepl) as well.

**Comment by DamianEdwards** [comment, marked as answer]
This is now available in .NET 10 preview 4: https://www.youtube.com/watch?v=98MizuB7i-w

**Reply by TahirAhmadov** [reply]
This is such an important addition. Kudos to the team for working on this. C# is a great language and needs to be easy to teach to beginners. Keep them coming 👍

**Comment by The-Futurist** [comment]
Very helpful stuff.
